File: Team 3/Typescript/Regular Expression Matching.ts
function match(s: string, p: string): void {
    if (p[0] === '*') {
        console.log(false);
        return;
    }

    // Reducción de asteriscos consecutivos
    let pArr: string[] = p.split('');
    let i: number = 0;
    while (i < pArr.length - 1) {
        if (pArr[i] === '' && pArr[i + 1] === '') {
            pArr.splice(i + 1, 1);
        } else {
            i++;
        }
    }
    p = pArr.join('');

    let sIndex: number = 0;
    let pIndex: number = 0;
    let count: number = 0;
    let aux1: string[][] = [[]];  // matriz auxiliar 
    let aux2: string[][] = [[]];

    while (sIndex < s.length || pIndex < p.length) {
        if (pIndex < p.length && sIndex < s.length && (s[sIndex] === p[pIndex] || p[pIndex] === '.')) {
            let sublistsToRemove: string[][] = [];
            aux1.forEach(sublist => {
                if (p[pIndex] === '.') {
                    if (sublist.length < s.length) {
                        sublist.push(s[sublist.length]);
                    }
                } else {
                    sublist.push(s[sIndex]);
                }

                if (sublist.join('') !== s.slice(0, sublist.length)) {
                    sublistsToRemove.push(sublist);
                }
            });

            sublistsToRemove.forEach(sublist => {
                let index = aux1.indexOf(sublist);
                if (index > -1) {
                    aux1.splice(index, 1);
                }
            });
            pIndex++;
            sIndex++;
            continue;
        }

        if (pIndex < p.length && p[pIndex] === '*') {
            if (sIndex >= s.length) {
                sIndex--;
            }

            if (pIndex > 0 && p[pIndex - 1] === '.') {
                aux2 = deepCopy(aux1);
                aux2.forEach(sublist => {
                    sublist.push(s[sIndex]);
                    while (sublist.join('') === s.slice(0, sublist.length)) {
                        aux1.push([...sublist]);
                        sublist.push(s[sIndex]);
                        if (sublist.length > s.length) {
                            break;
                        }
                    }
                });
            } else {
                aux2 = deepCopy(aux1);
                aux2.forEach(sublist => {
                    while (sublist.join('') === s.slice(0, sublist.length)) {
                        aux1.push([...sublist]);
                        sublist.push(sublist[sublist.length - 1]);
                        if (sublist.length > s.length) {
                            break;
                        }
                    }
                });
            }
            pIndex++;
        } else {
            if (pIndex >= p.length) {
                break;
            }

            aux1.forEach(sublist => {
                sublist.push(p[pIndex]);
            });

            aux1.forEach(sublist => {
                if (sublist.join('') === s && pIndex === p.length - 1) {
                    console.log(true);
                    return;
                } else if (sublist.join('') !== s && pIndex === p.length - 1) {
                    count++;
                }
            });

            if (count === aux1.length) {
                console.log(false);
                return;
            } else {
                count = 0;
            }

            if (aux1.length === 0) {
                break;
            }

            pIndex++;
            sIndex++;
        }
    }

    aux1.forEach(sublist => {
        if (sublist.join('') === s) {
            console.log(true);
            return;
        }
    });

    console.log(false);
}

function deepCopy(original: string[][]): string[][] {
    return original.map(sublist => [...sublist]);
}

File: Team 3/Python/Regular Expression Matching.py
import copy

def match(s, p):
  
    if p[0] == "*":
        print(False)
        return

    i = 0
    while i < len(p) - 1:
        if p[i] == "" and p[i+1] == "":
            p = p[:i] + p[i+1:]
        else:
            i += 1

    i = 0  
    j = 0  
    count = 0
    aux1 = [[]]  
    aux2 = [[]]  

    while i < len(s) or j < len(p):  
        if j < len(p) and (i < len(s) and (s[i] == p[j] or p[j] == ".")):
            
            sublists_to_remove = []
            for sublist in aux1:
                
                if p[j] == ".":
                    if len(sublist) < len(s):  
                        sublist.append(s[len(sublist)])
                else:
                    sublist.append(s[i])
                
                if ''.join(sublist) != s[:len(sublist)]:
                    sublists_to_remove.append(sublist)
            for sublist in sublists_to_remove:
                aux1.remove(sublist)  
            j += 1
            i += 1  
            continue
        
        if j < len(p) and p[j] == "*":
            if i >= len(s):
                i -= 1
            if j > 0 and p[j - 1] == ".":  
                aux2 = copy.deepcopy(aux1)
                for sublist in aux2:

                    sublist.append(s[i])  
                    while ''.join(sublist) == s[:len(sublist)]:
                        aux1.append(sublist.copy())  
                        sublist.append(s[i])
                        if len(sublist) > len(s):
                            break
            else:
                aux2 = copy.deepcopy(aux1)
                for sublist in aux2:
                    while ''.join(sublist) == s[:len(sublist)]:
                        aux1.append(sublist.copy())
                        sublist.append(sublist[-1])
                        if len(sublist) > len(s):
                            break
            j += 1  
        

        else:
            if j >= len(p):
                break
            for sublist in aux1:
                sublist.append(p[j])
            for sublist in aux1:
                if ''.join(sublist) == s and j == len(p) - 1:
                    print(True)
                    return
                elif ''.join(sublist) != s and j == len(p) - 1:
                    count += 1
            if count == len(aux1):
                print(False)
                return
            else:
                count = 0
            if len(aux1) == 0:
                break
            
            j += 1
            i += 1
        
    for sublist in aux1:
        if ''.join(sublist) == s:
            print(True)
            return
    
    print(False)